src/java/yaw/engine/items/ItemGroup.java

```diff
@@ -1,154 +1,210 @@
-        mItems.add(i);
-        i.addToGroup(this);
+    public void add(String s, ItemObject i) {
+        float nWeight = weight + i.getMesh().getWeight();
+        double ratioGr = weight / (double) nWeight, ratioI = i.getMesh().getWeight() / (double) nWeight;
+        weight = nWeight;
+        //center = new Vector3f((float) ((center.x * ratioGr) + (i.getPosition().x * ratioI)), (float) ((center.y * ratioGr) + (i.getPosition().y * ratioI)), (float) ((center.z * ratioGr) + (i.getPosition().z * ratioI)));
+        items.put(s, i);
+        updateCenter();
     }
 
-    /**
-     * remove an item of the group and update center
+
+    /** remove the item i from the group, update the center and the weight
+     *
+     * @param i the item to remove
      */
-    public void remove(Item i) {
-        mItems.remove(i);
-        i.removeFromGroup(this);
-        this.mWeight -= i.getAppearance().getWeight();
+    public void remove(ItemObject i) {
+        items.remove(i);
+        this.weight -= i.getMesh().getWeight();
         updateCenter();
     }
 
-    public void updateCenter() {
+
+    /**
+     * Update the center of the group
+     */
+    public void updateCenter(){
         double x = 0, y = 0, z = 0;
-        for (Item i : mItems) {
-            int w = i.getAppearance().getWeight();
-            x += i.getPosition().x * w;
-            y += i.getPosition().y * w;
-            z += i.getPosition().z * w;
+        for (String s  : items.keySet()) {
+            Item tmp = items.get(s);
+            int w = items.get(s).getMesh().getWeight();
+            x += tmp.getPosition().x ;
+            y += tmp.getPosition().y ;
+            z += tmp.getPosition().z ;
+
         }
-        x /= mWeight;
-        y /= mWeight;
-        z /= mWeight;
+        x /= (items.size()*1.f);
+        y /= (items.size()*1.f);
+        z /= (items.size()*1.f);
         mCenter = new Vector3f((float) x, (float) y, (float) z);
     }
 
-    /**
-     * Apply translation for all items in the group
-     */
-    public void translate(Vector3f translation) {
-        for (Item i : mItems)
-            i.getPosition().add(translation);
+    @Override
+    public void rotate(float x, float y, float z) {
+
+        Vector3f rotation = new Vector3f(x, y, z);
+        for (String s  : items.keySet()){
+            items.get(s).getRotation().add(rotation);
+            items.get(s).revolveAround(mCenter, rotation.x, rotation.y, rotation.z);
+
+        }
+
     }
 
-    /**
-     * Apply translation for all items in the group
-     */
+    @Override
     public void translate(float x, float y, float z) {
         Vector3f translation = new Vector3f(x, y, z);
-        for (Item i : mItems)
-            i.getPosition().add(translation);
+        for (String s  : items.keySet()){
+            items.get(s).getPosition().add(translation);
```

Comment:
Don't use think it's not better to iterate on the  `MapEntry`  objects instead ? Seems like a code smell here...